src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+      } else {
+        // FIXME: at this point we should set a flag to recover
+        // this inode in a /_recovery/<inodeno>.data file as we
+        // can't recover it into its desired filesystem position.
+        derr << "Dentry 0x" << std::hex
+          << parent_ino << std::dec << "/"
+          << dname << " already exists but points to 0x"
+          << std::hex << existing_dentry.inode.ino << std::dec << dendl;
+        // Fall back to lost+found!
+        break;
+      }
+    }
+
+    // Inject linkage
+    // ==============
+    if (write_dentry) {
+      InodeStore dentry;
+      if (i == backtrace.ancestors.begin()) {
+        // This is the linkage for a file
+        dentry.inode.mode = 0500 | S_IFREG;
+        dout(10) << "Linking inode 0x" << std::hex << ino
+          << " at 0x" << parent_ino << "/" << dname << std::dec
+          << " with size=" << file_size << " bytes" << dendl;
+
+        // The file size and mtime we learned by scanning globally
+        dentry.inode.size = file_size;
+        dentry.inode.max_size_ever = file_size;
+        dentry.inode.mtime.tv.tv_sec = file_mtime;
+        dentry.inode.atime.tv.tv_sec = file_mtime;
+        dentry.inode.ctime.tv.tv_sec = file_mtime;
+
+        dentry.inode.layout = g_default_file_layout;
+        dentry.inode.layout.fl_object_size = chunk_size;
+        dentry.inode.layout.fl_stripe_unit = chunk_size;
+        dentry.inode.layout.fl_pg_pool = data_pool_id;
+
+        dentry.inode.truncate_seq = 1;
+        dentry.inode.truncate_size = -1ull;
+
+        dentry.inode.inline_data.version = CEPH_INLINE_NONE;
+      } else {
+        // This is the linkage for a directory
+        dentry.inode.mode = 0755 | S_IFDIR;
+
+        // Set nfiles to something non-zero, to fool any other code
+        // that tries to ignore 'empty' directories.  This won't be
+        // accurate, but it should avoid functional issues.
+        dentry.inode.dirstat.nfiles = 1;
+        dentry.inode.size = 1;
+
+      }
+      dentry.inode.nlink = 1;
+      dentry.inode.ino = ino;
+      dentry.inode.version = 1;
+      dentry.inode.backtrace_version = 1;
+      r = inject_linkage(parent_ino, dname, fragment, dentry);
+      if (r < 0) {
+        return r;
+      }
+    }
+
+    if (!created_dirfrag) {
+      // If the parent dirfrag already existed, then stop traversing the
+      // backtrace: assume that the other ancestors already exist too.  This
+      // is an assumption rather than a truth, but it's a convenient way
+      // to avoid the risk of creating multiply-linked directories while
+      // injecting data.  If there are in fact missing ancestors, this
+      // should be fixed up using a separate tool scanning the metadata
+      // pool.
+      break;
+    } else {
+      // Proceed up the backtrace, creating parents
+      ino = parent_ino;
+    }
+
+    // TODO ensure that injected inode has layout pointing to this pool (if
+    // none of its ancestor layouts does)
```

Comment:
Hmm, we should also use these when trying to determine the file layout. If a convenient ancestor has a layout that matches, that's probably the one, huh?